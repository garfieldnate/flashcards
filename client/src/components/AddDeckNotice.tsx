import { Text } from 'native-base';
import React, { Component } from 'react';
import { StyleSheet, View } from 'react-native';

import AddDeckButton from './AddDeckButton';

interface IProps {
  onPress: () => void;
}

export default class AddDeckNotice extends Component<IProps> {
  public render() {
    // using flex to center vertically
    return (
      <View style={styles.container}>
        <View style={{ flex: 1 }} />
        <View style={{ flex: 1 }}>
          <View style={styles.row}>
            <Text style={styles.text}>Touch</Text>
          </View>
          <View style={styles.row}>
            <AddDeckButton
              iconStyle={styles.icon}
              onPress={this.props.onPress}
            />
          </View>
          <View style={styles.row}>
            <Text style={styles.text}>to add a deck</Text>
          </View>
        </View>
        <View style={{ flex: 1 }} />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    flex: 1,
    justifyContent: 'center',
  },
  icon: {
    fontSize: 36,
  },
  row: {
    alignItems: 'center',
  },
  text: {
    fontSize: 24,
  },
});
