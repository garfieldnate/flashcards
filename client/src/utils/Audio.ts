import { Audio } from 'expo-av';
import { Sound } from 'expo-av/build/Audio';
import { PlaybackSource } from 'expo-av/build/AV';

export const loadAudio = async (src: PlaybackSource): Promise<Sound> => {
  const sound = new Audio.Sound();
  await sound.loadAsync(src);
  return sound;
};
