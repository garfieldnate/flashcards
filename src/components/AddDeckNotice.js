import React, { Component } from 'react';
import { StyleSheet, View } from 'react-native';
import { Body, Row, Text } from 'native-base';

import AddDeckButton from './AddDeckButton.tsx';

export default class AddDeckNotice extends Component {
  render() {
    // using flex to center vertically
    return (
        <View style={styles.container}>
            <View style={{flex: 1}} />
            <View style={{flex: 1}}>
                <View style={styles.row}>
                  <Text style={styles.text}>Touch</Text>
                </View>
                <View style={styles.row}>
                  <AddDeckButton iconStyle={styles.icon} navigation={this.props.navigation} />
                </View>
                <View style={styles.row}>
                  <Text style={styles.text}>to add a deck</Text>
                </View>
            </View>
            <View style={{flex: 1}} />
        </View>);
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center'
  },
  row: {
    alignItems: 'center'
  },
  icon: {
    fontSize: 36
  },
  text: {
    fontSize: 24
  }
});
