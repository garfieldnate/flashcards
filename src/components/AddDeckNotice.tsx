import React, { Component } from 'react';
import { StyleSheet, View } from 'react-native';
import { Body, Row, Text } from 'native-base';
import { NavigationScreenProp, NavigationState, NavigationParams } from 'react-navigation';

import AddDeckButton from './AddDeckButton';

type Navigation = NavigationScreenProp<NavigationState, NavigationParams>;

type Props = {
  navigation: Navigation,
};

export default class AddDeckNotice extends Component<Props> {
  render() {
    // using flex to center vertically
    return (
      <View style={styles.container}>
          <View style={{ flex: 1 }} />
          <View style={{ flex: 1 }}>
              <View style={styles.row}>
                <Text style={styles.text}>Touch</Text>
              </View>
              <View style={styles.row}>
                <AddDeckButton iconStyle={styles.icon} navigation={this.props.navigation} />
              </View>
              <View style={styles.row}>
                <Text style={styles.text}>to add a deck</Text>
              </View>
          </View>
          <View style={{ flex: 1 }} />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  row: {
    alignItems: 'center',
  },
  icon: {
    fontSize: 36,
  },
  text: {
    fontSize: 24,
  },
});
