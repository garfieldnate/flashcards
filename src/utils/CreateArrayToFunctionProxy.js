// the swiper library takes an array of cards and stores an internal pointer to
// the current card. Since we are generating cards as we go, this is really
// difficult to work with. To get around it, we create a proxy that pretends
// to be an array but underneath calls StudyManager.getNextCard().
const createArrayToFunctionProxy = (studyManager, stackSize) => {
  var generatedCards = [];
  var length;

  // look ahead stackSize cards so that we can properly report the array
  // length when it matters to the swiper component
  const addBufferCards = (index) => {
    const bufferNeeded = Math.max(0, index - generatedCards.length + stackSize);
    console.log("generatedCards.length: " + generatedCards.length)
    console.log("index: " + index)
    console.log("stackSize: " + stackSize)
    console.log("adding buffer: " + bufferNeeded)
    for (var i = 0; i < bufferNeeded; i++) {
      var newCard = studyManager.getNextCard();
      if (newCard) {
        generatedCards.push(newCard);
      } else {
        length = generatedCards.length;
        break;
      }
    }
  }
  addBufferCards(0);

  var handler = {
    get: function(target, name) {
      console.log("proxying: " + name);
      if (name === 'length') {
        console.log("returning length: " + length)
        return length;
      } else if(isNaN(name)) {
        return;
      }
      addBufferCards(name);

      if (name < generatedCards.length) {
        console.log("returning previous: " + generatedCards[name])
        return generatedCards[name];
      } else {
        var newCard = studyManager.getNextCard();
        if (newCard) {
          generatedCards.push(newCard);
          console.log("returning new: " + newCard)
          return newCard;
        }
      }
      length = generatedCards.length;
      console.log("set length to " + length)
      console.log("returning undefined")
      return;
    }
  };
  return new Proxy({}, handler);
}

export default createArrayToFunctionProxy;
