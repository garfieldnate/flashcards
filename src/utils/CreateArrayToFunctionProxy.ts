import StudyManager from '../logic/StudyManager';

// the swiper library takes an array of cards and stores an internal pointer to
// the current card. Since we are generating cards as we go, this is really
// difficult to work with. To get around it, we create a proxy that pretends
// to be an array but underneath calls StudyManager.getNextCard().
const createArrayToFunctionProxy = (studyManager: StudyManager, stackSize: number) => {
  const generatedCards = [];
  let length: number;

  // look ahead stackSize cards so that we can properly report the array
  // length when it matters to the swiper component
  const addBufferCards = (index: number) => {
    const bufferNeeded = Math.max(0, index - generatedCards.length + stackSize);
    // console.log("generatedCards.length: " + generatedCards.length)
    // console.log("index: " + index)
    // console.log("stackSize: " + stackSize)
    // console.log("adding buffer: " + bufferNeeded)
    for (let i = 0; i < bufferNeeded; i += 1) {
      const newCard = studyManager.getNextCard();
      if (newCard) {
        generatedCards.push(newCard);
      } else {
        length = generatedCards.length;
        break;
      }
    }
  };
  addBufferCards(0);

  const handler = {
    get(target: any, name: 'length' | number) {
      // console.log("proxying: " + name);
      if (name === 'length') {
        // console.log("returning length: " + length)
        return length;
      } if (isNaN(name)) {
        return;
      }
      addBufferCards(name);

      if (name < generatedCards.length) {
        // console.log("returning previous: " + generatedCards[name])
        return generatedCards[name];
      }
      const newCard = studyManager.getNextCard();
      if (newCard) {
        generatedCards.push(newCard);
        // console.log("returning new: " + newCard)
        return newCard;
      }

      length = generatedCards.length;
      // console.log("set length to " + length)
      // console.log("returning undefined")
      return;
    },
  };
  return new Proxy<{}>({}, handler);
};

export default createArrayToFunctionProxy;
