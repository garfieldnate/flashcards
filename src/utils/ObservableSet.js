// Taken from https://github.com/mobxjs/mobx/issues/69#issuecomment-293256272
import {observable, ObservableMap } from 'mobx';
const { enhancer: deepEnhancer } = observable.map();


export default class ObservableSet extends ObservableMap {
  constructor(initialValues=[], name) {
    super(initialValues.map(v => [v,v]), deepEnhancer, name);
  }

  static create() {
    return new this(...arguments);
  }

  add(value) {
    return this.set(value, value);
  }

  forEach(callback, thisArg = this) {
    for(const value of this) {
      callback.call(thisArg, value);
    }
  }

  [Symbol.iterator]() {
    return this.keys();
  }
}
