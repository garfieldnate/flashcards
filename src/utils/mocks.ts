import { NavigationScreenProp, NavigationState, NavigationParams } from 'react-navigation';

type Navigation = NavigationScreenProp<NavigationState, NavigationParams>;
const mockNavigator = (): Navigation => {
  const mockNavigator: any = {
    navigate: jest.fn(),
    state: {
      params: {
        userData: 'foo',
        deckSource: 'bar',
      },
    },
  };
  return mockNavigator;
};

export { mockNavigator };
