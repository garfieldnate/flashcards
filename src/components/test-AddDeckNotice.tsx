import React from 'react';
import renderer from 'react-test-renderer';
import { mockNavigator } from '../utils/mocks';
import AddDeckNotice from './AddDeckNotice';

import { Ionicons } from '@expo/vector-icons';
import * as Font from 'expo-font';
beforeAll(() => {
  return Font.loadAsync(Ionicons.font);
});

it('renders icon button', () => {
  const tree = renderer
    .create(<AddDeckNotice navigation={mockNavigator()} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
