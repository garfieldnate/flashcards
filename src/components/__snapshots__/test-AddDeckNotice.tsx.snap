// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders icon button 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  />
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 16,
            },
            Object {
              "fontSize": 24,
            },
          ]
        }
        uppercase={false}
      >
        Touch
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
        }
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "center",
            "backgroundColor": "transparent",
            "borderBottomWidth": null,
            "borderColor": "#007aff",
            "borderLeftWidth": null,
            "borderRadius": 5,
            "borderRightWidth": null,
            "borderTopWidth": null,
            "borderWidth": undefined,
            "elevation": 0,
            "flexDirection": "row",
            "height": 45,
            "justifyContent": "space-between",
            "opacity": 1,
            "paddingBottom": 0,
            "paddingTop": 0,
            "shadowColor": null,
            "shadowOffset": null,
            "shadowOpacity": null,
            "shadowRadius": null,
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 12,
              },
              Array [
                Object {
                  "color": "#000",
                  "fontSize": 30,
                },
                Object {
                  "color": "#007aff",
                  "fontSize": 24,
                  "marginLeft": 16,
                  "marginRight": 16,
                  "paddingTop": 2,
                },
                Object {
                  "fontSize": 36,
                },
              ],
              Object {
                "fontFamily": "ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          ï„€
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 16,
            },
            Object {
              "fontSize": 24,
            },
          ]
        }
        uppercase={false}
      >
        to add a deck
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  />
</View>
`;
